pkg/health/membergarbagecollector/membergarbagecollector.go

```diff
@@ -0,0 +1,175 @@
+		}
+	}
+	namespace, err := utils.GetEnvVarOrError(podNamespace)
+	if err != nil {
+		logger.Fatalf("POD_NAMESPACE env var not present: %v", err)
+	}
+	podName, err := utils.GetEnvVarOrError(podName)
+	if err != nil {
+		logger.Fatalf("POD_NAME env var not present: %v", err)
+	}
+	return &MemberGarbageCollector{
+		logger:                logger.WithField("actor", "etcd-membergc"),
+		k8sClient:             clientSet,
+		podNamespace:          namespace,
+		stsName:               podName[:strings.LastIndex(podName, "-")],
+		etcdConnectionTimeout: etcdTimeout,
+	}, nil
+}
+
+// RunMemberGarbageCollectorPeriodically periodically calls the member garbage collector
+func RunMemberGarbageCollectorPeriodically(ctx context.Context, hconfig *brtypes.HealthConfig, logger *logrus.Entry, etcdConfig *brtypes.EtcdConnectionConfig) {
+	clientSet, err := miscellaneous.GetKubernetesClientSetOrError()
+	if err != nil {
+		logger.Errorf("failed to create kubernetes clientset: %v", err)
+		return
+	}
+	mgc, err := NewMemberGarbageCollector(logger, clientSet, etcdConfig.ConnectionTimeout)
+	if err != nil {
+		logger.Errorf("failed to initialize new etcd member garbage collector: %v", err)
+		return
+	}
+	mgc.gcTimer = time.NewTimer(hconfig.MemberGCDuration.Duration)
+	defer mgc.gcTimer.Stop()
+	mgc.logger.Info("Started etcd member garbage collector")
+
+	//Create etcd client to get etcd cluster member list
+	clientFactory := etcdutil.NewFactory(*etcdConfig)
+	etcdCluster, err := clientFactory.NewCluster()
+	if err != nil {
+		mgc.logger.Errorf("Failed to create etcd cluster client: %v", err)
+	}
+	defer etcdCluster.Close()
+
+	for {
+		select {
+		case <-mgc.gcTimer.C:
+			err := mgc.RemoveSuperfluousMembers(ctx, etcdCluster)
+			if err != nil {
+				mgc.logger.Warn(err)
+			}
+			mgc.gcTimer.Reset(hconfig.MemberGCDuration.Duration)
+		case <-ctx.Done():
+			mgc.logger.Info("Stopping etcd member garbage collector")
+			return
+		}
+	}
+}
+
+// RemoveSuperfluousMembers removes a member from the etcd cluster if its corresponding pod is not present
+func (mgc *MemberGarbageCollector) RemoveSuperfluousMembers(ctx context.Context, etcdCluster etcdclient.ClusterCloser) error {
+	sts := &appsv1.StatefulSet{}
+	if err := mgc.k8sClient.Get(ctx, client.ObjectKey{
+		Namespace: mgc.podNamespace,
+		Name:      mgc.stsName,
+	}, sts); err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Could not fetch etcd statefulset %v with error: %v", mgc.stsName, err),
+		}
+	}
+	stsSpecReplicas := *sts.Spec.Replicas
+	stsStatusReplicas := sts.Status.Replicas
+
+	listCtx, listCtxCancel := context.WithTimeout(ctx, mgc.etcdConnectionTimeout.Duration)
+	defer listCtxCancel()
+	etcdMemberListResponse, err := etcdCluster.MemberList(listCtx)
+	if err != nil {
+		return &errors.EtcdError{
+			Message: fmt.Sprintf("Could not list etcd cluster members: %v", err),
+		}
+	}
+
+	if stsSpecReplicas >= stsStatusReplicas && stsSpecReplicas >= int32(len(etcdMemberListResponse.Members)) {
```

Comment:
```suggestion
	if sts.generation != sts.status.observedGeneration && stsSpecReplicas != sts.status.updatedReplicas && stsSpecReplicas >= int32(len(etcdMemberListResponse.Members)) {
```

Considering `sts.status.updatedReplicas` gives us a more confidence that no update is happening. WDYT?

Reply:
Yes, makes sense. Done